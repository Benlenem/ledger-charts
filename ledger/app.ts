///<reference path="typings/index.d.ts" />

    /**
    * @description determine if an array contains one or more items from another array.
    * @param {array} haystack the array to search.
    * @param {array} arr the array providing items to check for in the haystack.
    * @return {boolean} true|false if haystack contains at least one item from arr.
    */
    function findOne(haystack: Array<any>, arr: Array<any>) {
        return arr.some(function (v) {
            return haystack.indexOf(v) >= 0;
        });
    };

    class Period {
        private _startDate: moment.Moment;
        private _endDate: moment.Moment;
        private _stats: any;

        constructor(startDate: moment.Moment, endDate: moment.Moment) {
            this._startDate = startDate; //Start date is inclusive
            this._endDate = endDate; //End date is exclusive
            this._stats = new Map<string, number>();
        }
        get startDate(): moment.Moment { return this._startDate; }
        get endDate(): moment.Moment { return this._endDate; }
        get stats(): Map<string, number> { return this._stats; }

        getName() {
            return this._startDate.format("DD/MM/YYYY") + " - " + this._endDate.format("DD/MM/YYYY");
        }
    }

    interface Transaction {
        header: Header;
        postings: Array<Posting>;
    }

    interface Header {
        date: string;
        title: string;
    }

    interface Posting {
        account: string;
        currency: Currency;
    }

    interface Currency {
        amount: number;
    }

    enum GroupBy { Account = 1, Year, Semester, Trimester, Month, Week, Day }
    enum StatParam { Sum = 1, Average }
    enum PeriodGap { None = 1, Year, Month, Week, Day }
    enum TransactionType { DEBT = 1, CREDIT, BOTH }

    class GoogleChartsVisu {
        drawPeriod(period: Period) {
            // Create the data table.
            let data = new google.visualization.DataTable();
            data.addColumn('string', 'Compte');
            data.addColumn('number', 'Total');

            Object.keys(period.stats).forEach(k => {
                let row: Array<google.visualization.ICell> = [];
                row.push({ v: k, f: null, p: null });
                row.push({ v: period.stats[k], f: null, p: null });

                data.addRow(row);
            });

            /*
            let title =
                this._sourceAccounts + ", "
                //      + fromDateInclusive + "-" + toDateInclusive
                + ", " + this._grouping
                + ", " + this._param;
            */

            // Set chart options

            var options = {
                /*'title': title,*/
                'pieSliceText': 'value',
                'width': 1000,
                'height': 300
            };


            let wrapper = new google.visualization.ChartWrapper({
                chartType: 'ColumnChart',
                dataTable: data,
                options: options,
                containerId: 'chart_div'
            });
            wrapper.draw(document.getElementById('chart_div'));

        }

        drawPeriods(periods) {
            var data = new google.visualization.DataTable();

            var indexes = new Set();

            periods.forEach(p => {
                Object.keys(p.stats).forEach(k => {
                    indexes.add(k);
                });
            });

            data.addColumn('string', 'Période');

            indexes.forEach(index => {
                data.addColumn('number', index);
            });

            periods.forEach(p => {
                let rowValues = [p.getName()].concat(Array.from(indexes).map(index => p.stats[index]));
                data.addRow(rowValues);
            });

            var options = {
                //isStacked: 'relative',
                isStacked: true,
                //interpolateNulls: true,
                title: 'Company Performance',
                hAxis: { title: 'Year', titleTextStyle: { color: '#333' } },
                vAxis: { minValue: 0 },
                pointsVisible: true
            };

            let wrapper = new google.visualization.ChartWrapper({
                chartType: 'AreaChart',
                dataTable: data,
                options: options,
                containerId: 'chart_div'
            });
            wrapper.draw();
        }
    }

    class Engine {
        private _transactions: Array<Transaction> = []
        private _sourceAccounts: Array<string>; //Comptes sur lesquels on réalise les stats
        private _periods: Array<Period>;
        private _grouping = GroupBy.Account;
        private _param = StatParam.Sum;
        private _visu: GoogleChartsVisu = new GoogleChartsVisu();
        private _type: TransactionType;

        constructor(transactions: Array<Transaction>) {
            this._sourceAccounts = [];
            this._periods = [];
            this._transactions = transactions;

            this.addMissingAmount();
        }

        createPeriods(startDate, endDate, gap, gapMultiple) {
            let from = startDate.clone();
            let periods = [];

            while (from < endDate) {
                let to;

                if (gap == PeriodGap.Year) {
                    to = from.clone().add(gapMultiple, 'year');
                }
                else if (gap == PeriodGap.Month) {
                    to = from.clone().add(gapMultiple, 'month');
                }
                else if (gap == PeriodGap.Week) {
                    to = from.clone().add(gapMultiple, 'week');
                }
                else if (gap == PeriodGap.Day) {
                    to = from.clone().add(gapMultiple, 'day');
                }
                else {
                    to = endDate;
                }
                periods.push(new Period(from, moment.min(to, endDate)));

                from = to.clone();
            }

            return periods;
        }

        getAllAccounts(): Array<string> {
            let accounts: Set<string> = new Set();

            this._transactions.forEach(tr => {
                tr.postings.forEach(ps => {
                    accounts.add(ps.account);
                });
            });

            return Array.from(accounts).sort((a1, a2) => a1.localeCompare(a2));
        }

        addMissingAmount() {
            this._transactions.forEach(tr => {

                let totalSum: number = 0;
                let incompletePosting;

                tr.postings.forEach(ps => {
                    if (ps.currency.amount) {
                        totalSum += ps.currency.amount;
                    }
                    else {
                        incompletePosting = ps;
                    }
                });

                if (incompletePosting) {
                    incompletePosting.currency.amount = - totalSum;
                }
            });
        }

        analyzeTransactions(
            accounts: Array<string>,
            startDate: string,
            endDate: string,
            groupy: GroupBy,
            statParam: StatParam,
            periodGap: PeriodGap,
            numPeriods: number,
            transactionType: TransactionType) {
            this._sourceAccounts = accounts;
            this._periods = this.createPeriods(moment(startDate, "DD/MM/YYYY"), moment(endDate, "DD/MM/YYYY"), periodGap, numPeriods);
            this._grouping = groupy;
            this._param = statParam;
            this._type = transactionType;

            this._transactions.forEach(tr => this.analyzeTransaction(tr));

            if (this._periods.length == 1) {
                this._visu.drawPeriod(this._periods[0]);
            }
            else {
                this._visu.drawPeriods(this._periods);
            }
        }

        analyzeTransaction(tr: Transaction) {
            this._periods.forEach(
                period => {
                    let transactionDate: moment.Moment = moment(tr.header.date, "YYYY/MM/DD")
                    if (transactionDate >= period.startDate && transactionDate < period.endDate) {
                        let accountNames: Array<any> = tr.postings.map(p => p.account);

                        if (findOne(this._sourceAccounts, accountNames)) {
                            this._sourceAccounts.forEach(a => {
                                let posting = tr.postings.find(p => p.account == a);

                                if (posting) {
                                    let index: string;
                                    if (this._grouping == GroupBy.Account) {
                                        index = a;
                                    }
                                    else if (this._grouping == GroupBy.Year) {
                                        index = String(transactionDate.year());
                                    }
                                    else if (this._grouping == GroupBy.Semester) {

                                        index = String(Math.floor(transactionDate.month() / 6));
                                    }
                                    else if (this._grouping == GroupBy.Trimester) {
                                        index = String(Math.floor(transactionDate.month() / 3));
                                    }
                                    else if (this._grouping == GroupBy.Month) {
                                        index = String(transactionDate.month());
                                    }
                                    else if (this._grouping == GroupBy.Week) {
                                        index = String(transactionDate.week());
                                    }
                                    else if (this._grouping == GroupBy.Day) {
                                        index = String(transactionDate.weekday());
                                    }

                                    let amount: number = posting.currency.amount || 0;

                                    if (this._type == TransactionType.BOTH ||
                                        (amount < 0 && this._type == TransactionType.DEBT) ||
                                        (amount > 0 && this._type == TransactionType.CREDIT)) {
                                        
                                        let stats: Map<string, number> = period.stats;

                                        if (this._param == StatParam.Sum) {
                                            stats[index] = (stats[index] || 0) + amount;
                                        }
                                        else if (this._param == StatParam.Average) {
                                            stats[index] = stats[index] ? (stats[index] + amount) / 2 : amount;
                                        }
                                    }
                                }
                            });
                        }
                    }
                });
        }
    }


window.onload = () => {
    
};