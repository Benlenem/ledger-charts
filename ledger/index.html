<html content="text/html; charset=utf-8">

<head>
    <meta charset="UTF-8">
    <!--Load the AJAX API-->
    <script type="text/javascript" src="https://www.gstatic.com/charts/loader.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.14.1/moment.min.js"></script>
    <script type="text/javascript" src="http://requirejs.org/docs/release/2.3.1/minified/require.js"></script>
    <script type="text/javascript" src="./app.js"></script>
    <script type="text/javascript" src="./parser_ledger.js"></script>

</head>

<body>

    <input id="file" type="file" />
    <button onclick="javascript:loadTransactions();">Visualiser</button>
    <br/>

    <select multiple id="fromAccounts" style="height:150px;"></select>

    <!--Div that will hold the pie chart-->
    <div id="chart_div"></div>
    <span id="nb_transactions"/>
    <br/>
    
    <script>

        var transactions = [];
        let engine;

        window.onload = function () {
            // Load the Visualization API and the corechart package.
            google.charts.load('current', { 'packages': ['corechart'] });

            var fileInput = document.querySelector('#file');

            fileInput.addEventListener('change', function () {

                var reader = new FileReader();

                reader.addEventListener('load', function () {
                    //console.log("@" + reader.result + "@");
                    transactions = PARSER.parse(reader.result);
                  //  document.getElementById("nb_transactions").innerHTML = "Nb transaction : " + transactions.length;

                    engine = new Engine(transactions);

                    var x = document.getElementById("fromAccounts");
                    for(account of engine.getAllAccounts()) {
                        var option = document.createElement("option");
                        option.text = account;
                        x.add(option);
                    }
                });

                reader.readAsText(fileInput.files[0]);

            });

            
        }

        function loadTransactions() {
            var x = document.getElementById("fromAccounts");
            var accounts = getSelectedValues(x);
            console.log(accounts);
            engine.analyzeTransactions(accounts);
        }

        function getSelectedValues(select) {
            var result = [];
            var options = select && select.options;
            var opt;

            for (var i = 0, iLen = options.length; i < iLen; i++) {
                opt = options[i];

                if (opt.selected) {
                    result.push(opt.value || opt.text);
                }
            }
            return result;
        }
    </script>
</body>

</html>
