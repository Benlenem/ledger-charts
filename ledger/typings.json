{
  "globalDependencies": {
    "google-apps-script/google-apps-script.charts": "registry:dt/google-apps-script/google-apps-script.charts#0.0.0+20160316171810",
    "moment": "registry:dt/moment#2.11.1+20160829143156"
  }
}
