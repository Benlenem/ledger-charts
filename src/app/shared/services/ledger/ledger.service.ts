
import {from as observableFrom,  Observable } from 'rxjs';

import {toArray, map, mergeMap} from 'rxjs/operators';
import { Injectable } from '@angular/core';

import { Transaction } from '../../models/transaction';

import Decimal from 'decimal.js';
import * as moment from 'moment';
import * as pegjs from 'pegjs';


import { Posting } from '../../models/posting';



@Injectable()
export class LedgerService {

  private _parser: pegjs.Parser;

  constructor() {
    const grammar =
    `
    Start
    = NewLineOrCommentBlock? 
      trs:(
        tr:Transaction Newline?
          NewLineOrCommentBlock? {return tr}
      )*
      { return trs }
  
  NewLineOrCommentBlock = (Comment / Newline)*
      
  Transaction
    = fl:FirstLine
      sls:(sl:SecondLine { return sl} )+
      { return { header:fl, postings:sls }}
  
  FirstLine
    = d:Date _* tg:(tag:ClearingTag _ { return tag })? t:StringWithoutComment c:Comment? Newline
    {
      return { date:d, title:t.trim(), clearTag:tg, tags:c ? c.tags : [], comments: c ? c.str.trim() : null }
    }
  
  SecondLine
    =
    _+
    ct:(ct:ClearingTag _ { return ct })?
    a:Account
    _*
    c:CurrencyCombinations?
    _*
    cm:Comment?
    Newline?
    {
      return {
        tag:ct,
        account:a,
        sign:c ? c.sign : null,
        amount:c ? c.amount : null,
        currency:c ? c.currency : null,
        comment:cm ? cm.str : null,
        tags:cm ? cm.tags : []
      }
    }
  
  Comment
    = CommentStartChars _* str:StringWithTags { return str; }
    
  CommentStartChars = [';']
  
  CurrencyCombinations = a1:Line / a2:Line2 / a3:Line3
  
  Line = 
      sign:'-'? _* num:Number _* n:CurrencyName?
      {
        return { sign:sign, currency:n, amount:num }
      }
      
  Line2 = 
      sign:'-'? _* n:CurrencyName? _* num:Number
      {
        return { sign:sign, currency:n, amount:num }
      }
      
  Line3 =
    n:CurrencyName? _* sign:'-'? _* num:Number
      {
        return { sign:sign, currency:n, amount:num }
      }
      
  Number = $([0-9]+ ('.' decimals:[0-9]*)?)
    
  CurrencyName
    = '"' chars:CurrencyNameWithEscaping '"' { return chars.join("") }
    / chars:CurrencyNameWithoutEscaping { return chars.join("") }
  
  CurrencyNameWithoutEscaping
    = (!(CommentStartChars / Newline / "-" / "+" / '"' / [0-9] / _) c:. { return c })+
  
  CurrencyNameWithEscaping 
    = (!(CommentStartChars / Newline / "-" / "+" / '"') c:. { return c })+
  
  Date
    = d:((d:[0-9]+ {return d.join("")})'/'(m:[0-9]+ { return m.join("") })'/'(y:[0-9]+ { return y.join("")} )) 
    { return d.join("") }
  
  ClearingTag
    = ['*','!']
  
  StringWithoutComment
    = t:(
      !(CommentStartChars / Newline) c:. { return c }
      )+
      { return t.join('') }
  
  StringWithTags = rs:RawString {
    const t = rs.split(':');
      
    if(t.length > 2) { //there is a tag in it
      const tags = t.splice(1, t.length - 2);
      return { str:t.join('').trim(), tags }
    } else {
      return { str:rs.trim(), tags:[] } 
    }
  }
  
  RawString = chars:(!Newline c:. { return c })* { return chars.join("") }
  
  Account
    =  letters:(!"  " !"\\t" !"\\n" letter:. { return letter })* 
    {
      return letters.join("")
    }
  
  _ "whitespace or tab"
    = [' ', '\\t']
  
  Newline
  = '\\n' / '\\n\\r'/ '\\r\\n' / '\\r'
  `;
  
  this._parser = pegjs.generate(grammar);
  }

  parseLedgerString(text: string): Observable<Transaction[]> {
    return this.parseObservable(text).pipe(
    mergeMap(trs => observableFrom(trs)),
    map( t => {
      return {
        uuid: undefined,
        header : {
          date: moment.utc(t.header.date, 'YYYY/MM/DD').unix(),
          title: t.header.title,
          isVerified: t.header.clearTag === '*',
          tags: t.header.tags,
        },
        postings : t.postings.map(p => {
          const pt: Posting = {
            tag: p.tag,
            account: p.account,
            currency: p.currency,
            comment: p.comment,
            tags: p.tags,
          };

          pt.amount = p.amount ?  (p.sign || '') + p.amount : undefined;

          return pt;
        }),
      };
    }),
    toArray(),);
  }

  private parseObservable(text: string): Observable<LedgerTransaction[]> {
    return new Observable<LedgerTransaction[]>( obs => {
      obs.next(this._parser.parse(text) as LedgerTransaction[]);
      obs.complete();
    });
  }

  generateLedgerString(transactions: Transaction[]): Observable<string> {
    return new Observable( obs => {
      let out = '';
      transactions.forEach( tr => {

        out += [
          moment.unix(tr.header.date).format('YYYY/MM/DD'),
          tr.header.isVerified ? '*' : null,
          tr.header.title,
          tr.header.tags.length > 0 ? ';:' + tr.header.tags.join(':') + ':' : ''
        ]
          .filter(val => val)
          .join(' ') + '\n';


        tr.postings.forEach( p => {
            out += '    ';
            out += [p.tag, p.account].filter(val => val).join(' ');

            if (p.amount) {
              out += '    ';
              out += [p.currency, p.amount].filter(val => val).join(' ');
            }

            if (p.comment || (p.tags && p.tags.length > 0)) {
              out += ' ; ';
              
              if (p.comment) {
                out +=  p.comment;
              }
              if (p.tags && p.tags.length > 0) {
                out += ' :' + p.tags.join(':') + ':'
              }
            }

            out += '\n';
        });

        out += '\n';
      });

      obs.next(out);
      obs.complete();
    });
  }
}

export interface LedgerTransaction {
  header: LedgerHeader;
  postings: Array<LedgerPosting>;
}

type ClearTag = "*" | "!";

interface LedgerHeader {
  date: string;
  title: string;
  clearTag: ClearTag | undefined;
  tags: string[],
}

export interface LedgerPosting {
  tag: string | undefined;
  sign: string | undefined;
  account: string;
  amount: string | undefined;
  currency: string | undefined;
  comment: string | undefined;
  tags: string[],
}
