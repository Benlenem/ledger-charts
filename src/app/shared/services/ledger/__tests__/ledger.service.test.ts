import { emptyFile, oneTransactionFile, transactions } from './fixtures';
import { LedgerService } from '../ledger.service';
import 'rxjs/add/operator/toPromise';
import 'rxjs/add/observable/zip';
import 'rxjs/add/operator/mergeMap';
import 'rxjs/add/operator/do';

describe(LedgerService.name, () => {
  it('parses empty file', () => {
    const service = new LedgerService();

    return expect(service.parseLedgerString(emptyFile).toPromise()).resolves.toMatchSnapshot();
  });

  it('parses file with one transaction', () => {
    const service = new LedgerService();

    return expect(service.parseLedgerString(oneTransactionFile).toPromise()).resolves.toMatchSnapshot();
  });

  it('saves transactions', () => {
    const service = new LedgerService();
    return expect(service.generateLedgerString(transactions).toPromise()).resolves.toMatchSnapshot();
  });

  it('loads and saves transactions', () => {
    const service = new LedgerService();

    let parsedTransactions;
    const loadAndSave = service.parseLedgerString(oneTransactionFile)
      .do(x => parsedTransactions = x)
      .flatMap(tr => service.generateLedgerString(tr).flatMap(value => service.parseLedgerString(value)));

    return expect(loadAndSave.toPromise()).resolves.toEqual(parsedTransactions);
  });
});
