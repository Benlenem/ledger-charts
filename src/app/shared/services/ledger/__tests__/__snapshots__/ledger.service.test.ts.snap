// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LedgerService parses empty file 1`] = `Array []`;

exports[`LedgerService parses file with one transaction 1`] = `
Array [
  Object {
    "header": Object {
      "date": 1484265600,
      "tag": "!",
      "tags": Array [
        "TAG0",
      ],
      "title": "Transaction",
    },
    "postings": Array [
      Object {
        "account": "CompteA:Sous compte",
        "amount": "41.3",
        "comment": "Commentaire",
        "currency": "€",
        "tag": null,
        "tags": Array [
          "TAG 1",
          "TAG 2",
        ],
      },
      Object {
        "account": "CompteA:Autre compte",
        "amount": "11.3",
        "comment": "Commentaire",
        "currency": "€",
        "tag": null,
        "tags": Array [
          "TAG 3",
        ],
      },
      Object {
        "account": "CompteB",
        "amount": undefined,
        "comment": null,
        "currency": null,
        "tag": "*",
        "tags": Array [],
      },
    ],
    "uuid": undefined,
  },
]
`;

exports[`LedgerService saves transactions 1`] = `
"2013/02/08 Transaction title
    ! accountA    € 3.14 ; Comment
    accountB

"
`;
